### PR TITLE
Moved timing and scheduling logic into school_schedule.py.

### PR DESCRIPTION
**Pull request by wkbraid**
My main goal was to put all the date and time logic together, and make it test-able.

Main changes:
* Date and time logic now lives in school_schedule.py
   - I was able to outsource some of the complexity using `numpy.busday_count()`.
   - Registration is no longer counted as open weekends, I believe this is the desired behavior.
* Added tests in school_schedule_test.py, run them with the `pytest` command.

Minor, misc:
* Updated README.md and requirements.txt
* Added password_template.py, and renamed `password` -> `server_password`, `p` -> `email_password`.
* I wrote all my code using google style conventions

Style opinions:
* I included [docstrings](https://peps.python.org/pep-0257/), this can be good for two reasons:
   - Highlights "needed to use" information (other comments are more like "needed to develop")
    - Various tools will extract these strings to produce a high level summary, try running `pydoc school_schedule` in a terminal.
 * I used `_leading_underscores` to signal "internal" functions, i.e. helper functions for the developer of the file, but not intended to be used from elsewhere. Notice that they don't show up in `pydoc` and it's good practice *not* to test them (tests are meant to ensure that the "external" behavior is working as expected, internals are allowed/expected to change).
 * Switching to 'lower_case' naming was probably bad because it's inconsistent. You can/should pester me to change this.

Next steps I could take if you think this kind of refactor is helpful:
 * Separate out "app login" logic in a similar way.
 * Move RegistrationManager to its own file, and try to make all the logic test-able in the same way.
 
 Hope this helps!

### CODE REVIEW

**Comment by wkbraid** [comment]
Critiquing my code could include anywhere from opinions to questions to needed changes, here's a real example from my work recently with some comments from my colleagues:

image

File: school_schedule.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        return False
+
+    # Wednesday has a separate schedule
+    if query.strftime("%A") == "Wednesday":
+        return (OPEN_TIME <= query.time()) and (query.time() <= WEDNESDAY_CLOSE_TIME)
+    
+    return (OPEN_TIME <= query.time()) and (query.time() <= CLOSE_TIME)
+
+
+def free_period(query=datetime.now()):
```
Comment (shartse) [inline school_schedule.py]: It's not clear to me what kind of data this function returns - updating the comment to indicate that it's a member of `FREE_PATTERN` would help

File: school_schedule_test.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+class TestFreePeriod:
+    def test_first_day(self):
+        assert free_period(datetime(2022, 9, 7)) == "C"
+
+    def test_known_day(self):
+        assert free_period(datetime(2023, 3, 14)) == "G"
+
+
+class TestRegistrationOpen:
```
Comment (shartse) [inline school_schedule_test.py]: Currently, all these dates are sometime in the future or on the weekend. What happens if it's a valid day, at a valid time, just in the past? Would this be worth testing?

File: school_schedule_test.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from school_schedule import registration_open, free_period
+
+from datetime import datetime
+
+
+class TestFreePeriod:
```
Comment (shartse) [inline school_schedule_test.py]: I really like the way the tests are organized!

File: school_schedule.py (hunk 1 above)
Comment (wkbraid) [reply school_schedule.py]: Done.

**Pull request closed without merging**